src/libopensc/card-laser.c

```diff
@@ -0,0 +1,2746 @@
+
+	dh_session->ifd_y.value = malloc(SHA_DIGEST_LENGTH);
+	if (!dh_session->ifd_y.value)
+		LOG_ERROR_GOTO(ctx, rv = SC_ERROR_OUT_OF_MEMORY, "Cannot allocate private DH key");
+
+	RAND_bytes((unsigned char*)&rd, sizeof(rd));
+	SHA1((unsigned char *)(&rd), sizeof(rd), dh_session->ifd_y.value);
+	dh_session->ifd_y.len = SHA_DIGEST_LENGTH;
+	bn_ifd_y = BN_bin2bn(dh_session->ifd_y.value, dh_session->ifd_y.len, NULL);
+
+	dh = DH_new();
+	if (!dh)
+		LOG_ERROR_GOTO(ctx, rv = SC_ERROR_OUT_OF_MEMORY, "Cannot allocate DH key");
+
+	DH_set0_pqg(dh, bn_N, NULL, bn_g);	// dh->p, dh->g 
+	DH_set0_key(dh, NULL, bn_ifd_y);	// dh->priv_key
+
+	if (!DH_check(dh, &dh_check))
+		LOG_ERROR_GOTO(ctx, rv = SC_ERROR_INTERNAL, "OpenSSL 'DH-check' failed");
+	if (!DH_generate_key(dh))
+		LOG_ERROR_GOTO(ctx, rv = SC_ERROR_INTERNAL, "OpenSSL 'DH-generate-key' failed");
+
+	DH_get0_key(dh, &pub_key, NULL);
+
+	dh_session->ifd_p.value = (unsigned char *)OPENSSL_malloc(BN_num_bytes(pub_key));
+	if (!dh_session->ifd_p.value)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_OUT_OF_MEMORY);
+	dh_session->ifd_p.len = BN_bn2bin(pub_key, dh_session->ifd_p.value);
+
+	dh_session->shared_secret.value = (unsigned char *)OPENSSL_malloc(DH_size(dh));
+	if (!dh_session->shared_secret.value)
+		LOG_FUNC_RETURN(ctx, rv = SC_ERROR_OUT_OF_MEMORY);
+	
+#if !defined(LIBRESSL_VERSION_NUMBER)
+	dh_session->shared_secret.len = DH_compute_key_padded(dh_session->shared_secret.value, bn_icc_p, dh);
+#else
+	rv = _compute_key_padded(card, dh_session->shared_secret.value, DH_size(dh), bn_icc_p, bn_ifd_y, bn_N);
+	if (DH_size(dh) > rv) 
+		LOG_ERROR_GOTO(ctx, rv, "Failed to calculate shared secret");
+#endif
+	sc_log(ctx, "shared-secret(%"SC_FORMAT_LEN_SIZE_T"u) %s", dh_session->shared_secret.len,
+			sc_dump_hex(dh_session->shared_secret.value, dh_session->shared_secret.len));
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_4_SHORT, 0x86, 0x00, 0x00);
+	apdu.cla = 0x80;
+	apdu.lc = dh_session->ifd_p.len;
+	apdu.datalen = dh_session->ifd_p.len;
+	apdu.data = dh_session->ifd_p.value;
+	apdu.le = sizeof(dh_session->card_challenge);
+	apdu.resplen = sizeof(dh_session->card_challenge);
+	apdu.resp = dh_session->card_challenge;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_GOTO_ERR(ctx, rv, "APDU transmit failed");
+	rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_GOTO_ERR(ctx, rv, "'OPEN SM' failed");
+
+	memcpy(dh_session->session_enc, dh_session->shared_secret.value, sizeof(dh_session->session_enc));
+	memcpy(dh_session->session_mac, dh_session->shared_secret.value + 24, sizeof(dh_session->session_mac));
+	for (uu = 0; uu < sizeof(dh_session->session_enc); uu++) {
+		dh_session->session_enc[uu] ^= dh_session->card_challenge[uu];
+		dh_session->session_mac[uu] ^= dh_session->card_challenge[16 + uu];
+	}
+
+	sc_log(ctx, "session key enc: %s", sc_dump_hex(dh_session->session_enc, 16));
+	sc_log(ctx, "session key auth: %s", sc_dump_hex(dh_session->session_mac, 16));
```

Comment:
Maybe this is some information you don't want to leak to logs. If you still want to write this, please use the log level `SC_LOG_DEBUG_PIN`